packages/flutter/lib/src/widgets/visibility.dart

```diff
@@ -0,0 +1,209 @@
+  /// maintained if it cannot be recreated on demand. One example of when the
+  /// state would be maintained is if the child subtree contains a [Navigator],
+  /// since that widget maintains elaborate state that cannot be recreated on
+  /// the fly.
+  ///
+  /// If this is true, an [Offstage] widget is used to hide the child instead of
+  /// replacing it with [replacement].
+  final bool maintainState;
+
+  /// Whether to maintain animations within the [child] subtree when it is
+  /// not [visible].
+  ///
+  /// To set this, [maintainState] must also be set.
+  ///
+  /// Keeping animations active when the widget is not visible is even more
+  /// expensive than only maintaining the state.
+  ///
+  /// One example when this might be useful is if the subtree contains an
+  /// animated [Image] that must be kept synchronized with other animated images
+  /// elsewhere in the interface.
+  ///
+  /// If this is true, no [TickerMode] widget is used.
+  final bool maintainAnimation;
+
+  /// Whether to maintain space for where the widget would have been.
+  ///
+  /// To set this, [maintainAnimation] must also be set.
+  ///
+  /// Maintaining the size when the widget is not [visible] is roughly as
+  /// expensive as keeping animations running, and may in some circumstances be
+  /// cheaper if the subtree is simple and the [visible] property is frequently
+  /// toggled, since it avoids triggering a layout change when the [visible]
+  /// property is toggled. However, if the [child] subtree is not trivial then
+  /// it is significantly cheaper to not even keep the state (see
+  /// [maintainState]).
+  ///
+  /// If this is true, [Opacity] is used instead of [Offstage].
+  ///
+  /// See also:
+  ///
+  ///  * [AnimatedOpacity] and [FadeTransition], which apply animations to the
+  ///    opacity for a more subtle effect.
+  final bool maintainSize;
+
+  /// Whether to maintain the semantics for the widget when it is hidden (e.g.
+  /// for accessibility).
+  ///
+  /// To set this, [maintainSize] must also be set.
+  ///
+  /// By default, with [maintainSemantics] set to false, the [child] is not
+  /// visible to accessibility tools when it is hidden from the user. If this
+  /// flag is set to true, then accessibility tools will report the widget as if
+  /// it was present.
+  final bool maintainSemantics;
+
+  /// Whether to allow the widget to be interactive when hidden.
+  ///
+  /// To set this, [maintainSize] must also be set.
+  ///
+  /// By default, with [maintainInteractivity] set to false, touch events cannot
+  /// reach the [child] when it is hidden from the user. If this flag is set to
+  /// true, then touch events will nonetheless be passed through.
+  final bool maintainInteractivity;
+
+  @override
+  Widget build(BuildContext context) {
+    if (maintainSize) {
+      Widget result = child;
+      if (!maintainInteractivity) {
+        result = new IgnorePointer(
```

Comment:
Should we be concerned about maintaining the tree structure when this is toggled.
Specifically if I'm not wrong toggling maintainInteractivity might destroy the state of child widgets without a key... I can imagine it being confusing to have a Visible widget with maintainState equals true destroy your state when you toggle interactivity...

Reply:
Ohh I just saw your comment saying that the maintain flags should not be changed.
I'm a bit concerned that people (like I just did) might overlook that part in the comments and get confused by this.
I wonder if we can come up with an API where this will be more clear you're doing something wrong by changing the flags...?
One wild idea (which is probably bad) would be to have a separate widget class per maintain configuration....
Another option is maybe to make this a stateful widget just for the sake of asserting that the maintain flags does not change...?

Reply:
Another potential alternative: if maintainState is true wrap child with a keyed widget and allow to modify the maintain flags

Reply:
What would the use case be for changing the flags?

Reply:
Just the general approach that if it's there it will be used...